### ISSUE TITLE
u2oInterconnection breaks communication between pods when using hostport 

### ISSUE THREAD

**Issue by patriziobassi** [kubeovn/kube-ovn#2807]
# Expected Behavior

We have ngnix ingress controller cluster-wide deployed as container using hostPorts (and so using the workers' control plane IP address)

```
    ports:
    - containerPort: 80
      hostPort: 80
      name: http
      protocol: TCP
    - containerPort: 443
      hostPort: 443
      name: https
      protocol: TCP

```
those workers should reach underlay network in order to expose HTTP/HTTPs services using Service object and a HTTP backend similar to this:

```
# cat 01-pod.yml
kind: Pod
apiVersion: v1
metadata:
  name: apple-app
  labels:
    app: apple
  annotations:
    ovn.kubernetes.io/logical_switch: subnet1-vpc

spec:
  containers:
    - name: apple-app
      image: hashicorp/hashicorp/http-echo
      args:
        - "-text=apple"

---

kind: Service
apiVersion: v1
metadata:
  name: apple-service
spec:
  selector:
    app: apple
  ports:
    - port: 5678 # Default port for image

```

We are using a subnet definition for underlay network like this:

```
apiVersion: kubeovn.io/v1
kind: Subnet
metadata:
  name: subnet1-vpc
spec:
  cidrBlock: 10.0.0.0/24
  gateway: 10.0.0.1
  vlan: provvlan1
  protocol: IPv4
  excludeIps:
  - 10.0.0.1
  - 10.0.0.2..10.0.0.15
  - 10.0.0.32..10.0.0.255
  private: false
  gatewayType: distributed
  natOutgoing: false
  disableGatewayCheck: true
  u2oInterconnection: true

```

# Actual Behavior

We see nginx ingress to expose the service but when a client tries to connect it times out because the backend is unreachable, while we should get the traffic to the backend pod.

Changing 
` u2oInterconnection: false` solves this internal communication issue but of course breaks the underlay pod to Services communication

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-15T02:14:53Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"clean", BuildDate:"2023-04-15T02:05:42Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}
  ```

- kube-ovn version: 1.11.5

- operation-system/kernel version: "Ubuntu 22.04.2 LTS"

**Comment by patriziobassi** [comment]
Hi,
i can add more info: the problem is not specific to hostports of course but the reachablility of the underlay pod network from the control plane network.

Infact we first spot the problem when we had nginx deployed with hostport, so the load balancer could not reach the backend endpoints.

Today we tested metalLB in order to expose Services to external (non-k8s) clients: because of the routing the requests from the clients will go to the default gw and then from the control plane network should got to the pod network but they don't.
if we disable the u2o interconnection flag this works and pods can receive the forwarded traffic.

This is a show stopper for non http applications which need an external reachable ip.

we'll be happy to test a fix or a workaround
Thank you

**Comment by changluyi** [comment]
I can't reproduce your bug, in my kind env, you can provide more detailed information like below

env: 
1. kube-ovn: v1.11.5
2. kernel 3.10.0-1160.83.1.el7.x86_64

steps:
1. create kind cluster
```
make kind-init && make kind-install-underlay
```

2. install metallb
```
kubectl create namespace metallb-system
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.12.1/manifests/metallb.yaml
```

3. config external ippool
```configmap.yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: config
  namespace: metallb-system
data:
  config: |
    address-pools:
    - name: default
      protocol: layer2
      addresses:
      - 172.18.0.100-172.18.0.104

```
4. deployment and service
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:alpine
        ports:
        - name: http
          containerPort: 80

---
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  #loadBalancerIP: 172.18.0.100
  type: LoadBalancer
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
```

5. enable ovn-default u2oInterconection

result:
 # kubectl get svc
NAME         TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)        AGE
kubernetes   ClusterIP      10.96.0.1       <none>         443/TCP        51m
nginx        LoadBalancer   10.101.21.162   172.18.0.100   80:32601/TCP   67s

I can access 172.18.0.100: 80 successfully from external

**Comment by patriziobassi** [comment]
i think you did not deploy pods on the underlay network.
just discard the metallb thing:

1) deploy a cluster
2) create a subnet with the spec i posted above, ensure it has "  vlan: xxxx" and vlan is using a provider-network with a physical net interface (underlay)
3) deploy a pod in the underlay
4) from a the worker (just enter in ssh) you cannot reach the pod in the underlay

if you disable the interconnect it works.

**Comment by zhangzujian** [comment]
Could you please try to access the Pod by executing `curl -I <URL>` in the k8s node? @patriziobassi

**Comment by patriziobassi** [comment]
@zhangzujian when not using interconnect it works, when using it doesn't. i saw you linked something regarding MTUs but i don't think it's a MTU issue (unless that's any side effect...).

We did some tcpdump analisys and spotted this:

1) interconnect off:
when interconnect is off communication works because the pod has as default gw the real router, so when it replies it uses the gw mac address as destination.

2) interconnect on
with interconnect on pod have a virtual router so the packet has as destination this virtual mac address, than Kube-ovn replaces it with the real subnet gw. This is working fine when the communication *starts* from the pod (i.e. a curl from the port to an external address.

When from outside, from k8s worker node (even the control plane network with is on another L3 network, so it passed through the physical gw) we try to access a port exposes port we can see that the traffic enters the pod, it replies using its virtualrouter mac address and then the traffic is not going out to the physical interface.

we suspect we are missing a rule (conntrack like) that make the responses behave like when the initiate the traffic.

**Comment by changluyi** [comment]
> @zhangzujian when not using interconnect it works, when using it doesn't. i saw you linked something regarding MTUs but i don't think it's a MTU issue (unless that's any side effect...).
> 
> We did some tcpdump analisys and spotted this:
> 
> 1. interconnect off:
>    when interconnect is off communication works because the pod has as default gw the real router, so when it replies it uses the gw mac address as destination.
> 2. interconnect on
>    with interconnect on pod have a virtual router so the packet has as destination this virtual mac address, than Kube-ovn replaces it with the real subnet gw. This is working fine when the communication _starts_ from the pod (i.e. a curl from the port to an external address.
> 
> When from outside, from k8s worker node (even the control plane network with is on another L3 network, so it passed through the physical gw) we try to access a port exposes port we can see that the traffic enters the pod, it replies using its virtualrouter mac address and then the traffic is not going out to the physical interface.
> 
> we suspect we are missing a rule (conntrack like) that make the responses behave like when the initiate the traffic.

===
we try to access a port exposes port we can see that the traffic enters the pod, it replies using its virtualrouter mac address and then the traffic is not going out to the physical interface.

it will reroute to physical gw , you can see the policy route on  ovn-cluster, 
```
prio 29000 match: "ip4.src == underlay subnet cidr"                         action: reroute physical gw
```

**Comment by patriziobassi** [comment]
@changluyi that policy seems to work fine when i initiate the connection *from* the pod to the external IPs.
when it comes to replies we cannot see traffic con the physical port.

K8s worker has 10.0.5.40/24 address (ens3 interface)
Kubeovn subnet 10.0.8.0/24 - underlay with ens8 physical interface provider network

TEST:
1) deploy a pod "apple-app" in namespace  ns1
in the nb db we can see it
```

switch 42d05ba8-000c-47a0-9cfd-5f7c6e323196 (subnet1-vpc)
    port apple-app.ns1
        addresses: ["00:00:00:58:2A:EB 10.0.8.21"]
```

2) 
i issue a simple
`# curl 10.0.8.21:5678` 
from the K8s worker to the pod ip (as said, this works when not using u2o interconnection)


DEBUG:

so 00:00:00:58:2A:EB is the pod mac addr, 00:00:00:28:2d:9d is the u2o "virtual" router

Tracing inside the pod
kubectl ko tcpdump ns1/apple-app -lev port 5678

08:55:50.146465 <physical gw mac> > 00:00:00:58:2a:eb, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 63, id 47631, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.5.40.60232 > 10.0.8.21.5678: Flags [S], cksum 0xa9ec (correct), seq 2804690324, win 62720, options [mss 8860,sackOK,TS val 1669279569 ecr 0,nop,wscale 7], length 0
08:55:50.146505 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x547e), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718714520 ecr 1669279569,nop,wscale 7], length 0
08:55:51.149746 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x5093), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718715523 ecr 1669279569,nop,wscale 7], length 0
08:55:51.150352 <physical gw mac> > 00:00:00:58:2a:eb, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 63, id 47632, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.5.40.60232 > 10.0.8.21.5678: Flags [S], cksum 0xa5ff (correct), seq 2804690324, win 62720, options [mss 8860,sackOK,TS val 1669280574 ecr 0,nop,wscale 7], length 0
08:55:51.150371 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x5092), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718715524 ecr 1669279569,nop,wscale 7], length 0
08:55:53.165774 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x48b3), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718717539 ecr 1669279569,nop,wscale 7], length 0
08:55:53.166040 <physical gw mac> > 00:00:00:58:2a:eb, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 63, id 47633, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.5.40.60232 > 10.0.8.21.5678: Flags [S], cksum 0x9e1f (correct), seq 2804690324, win 62720, options [mss 8860,sackOK,TS val 1669282590 ecr 0,nop,wscale 7], length 0
08:55:53.166072 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x48b3), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718717539 ecr 1669279569,nop,wscale 7], length 0
08:55:57.201758 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x38ef), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718721575 ecr 1669279569,nop,wscale 7], length 0
08:56:05.389741 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x18f3), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718729763 ecr 1669279569,nop,wscale 7], length 0

we can clearly see that the pod replies using the 00:00:00:28:2d:9d router.

capturing in the worker node as root (on the wire)

tcpdump -i ens8 -lnev port 5678
tcpdump: listening on ens8, link-type EN10MB (Ethernet), snapshot length 262144 bytes
09:01:54.734843 <physical gw mac> > 00:00:00:58:2a:eb, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 63, id 21582, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.5.40.34156 > 10.0.8.21.5678: Flags [S], cksum 0x85c7 (correct), seq 3740135329, win 62720, options [mss 8860,sackOK,TS val 1669644158 ecr 0,nop,wscale 7], length 0
09:01:55.758045 <physical gw mac> > 00:00:00:58:2a:eb, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 63, id 21583, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.5.40.34156 > 10.0.8.21.5678: Flags [S], cksum 0x81c7 (correct), seq 3740135329, win 62720, options [mss 8860,sackOK,TS val 1669645182 ecr 0,nop,wscale 7], length 0
09:01:57.774040 <physical gw mac> > 00:00:00:58:2a:eb, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 63, id 21584, offset 0, flags [DF], proto TCP (6), length 60)
    10.0.5.40.34156 > 10.0.8.21.5678: Flags [S], cksum 0x79e7 (correct), seq 3740135329, win 62720, options [mss 8860,sackOK,TS val 1669647198 ecr 0,nop,wscale 7], length 0
	
Conclusion:
we can see only the requests from the physical gw from the pod, with the correct dst mac address, but we don't see the replies from the pod on the ens8 physical interface, while using the tcpdump inside the pod we see them.
So traffic is lost, or at least not correctly routed from the pod to the physical interface when replying

**Comment by zhangzujian** [comment]
Reply packets from the Pod go through ovn0, while ovn0's MTU is 1400.

**Comment by patriziobassi** [comment]
Ok! as it's just a 1 line patch i can try and rebuild the containers....in the meantime can you please explain me why is this affecting my case?

i see the SYN-ACK packets lost and they are quite small (74bytes) shouldn't a MTU of 1400 bytes be enough?

i saw the original bug the patch is targetting (https://github.com/kubeovn/kube-ovn/issues/2825) and the user had issues when transferring larger chunk (>2k bytes, which would make sense).

there are my MTUs, ovn0 is 8900 on the worker node

ip l|grep mtu
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
3: ens8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq master ovs-system state UP mode DEFAULT group default qlen 1000
4: ens9: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP mode DEFAULT group default qlen 1000
5: kube-ipvs0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default
6: ovs-system: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
7: mirror0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8900 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
8: ovn0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8900 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
9: genev_sys_6081: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65000 qdisc noqueue master ovs-system state UNKNOWN mode DEFAULT group default qlen 1000
10: br-int: <BROADCAST,MULTICAST> mtu 8900 qdisc noop state DOWN mode DEFAULT group default qlen 1000
11: br-provnet1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
13: 970643ede4ac_h@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8900 qdisc noqueue master ovs-system state UP mode DEFAULT group default qlen 1000

**Comment by changluyi** [comment]
I think it may be the mtu issue, the reply packet size more bigger then request, just have a try

**Comment by zhangzujian** [comment]
``` mermaid
graph LR

node-->ens3-->physical-gateway-->pod
```

``` mermaid
graph LR

pod-->logical-gateway-->ovn0-->node
```

If packets from Pod have a length greater than 9000, they will be dropped since these packets are sent with encapsulation and the encapsulated packets' length is greater than 9000.

> i see the SYN-ACK packets lost and they are quite small (74bytes) shouldn't a MTU of 1400 bytes be enough?

Are you capturing packets only on ens3? Capture packets on ovn0 and you will see them.

**Comment by changluyi** [comment]
when u2o enable, the reply packet doesn't pass through ovn0.

**Comment by patriziobassi** [comment]
actually my underlay pods are created with a default 1500 MTU

41: eth0@if42: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue

so this is not the problem. if you look at the trace the SYN-ACK (the reply) sent from pod to the worker is just 74bytes big

```
 10.0.8.21.5678 > 10.0.5.40.60232: Flags [S.], cksum 0x2351 (incorrect -> 0x5092), seq 2512576027, ack 2804690325, win 62636, options [mss 8960,sackOK,TS val 718715524 ecr 1669279569,nop,wscale 7], length 0
08:55:53.165774 00:00:00:58:2a:eb > 00:00:00:28:2d:9d, ethertype IPv4 (0x0800), length 74: (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60)
```

i did a tcpdump -i any on the worker node and i don't see traffic on ovn0 interface in this case.

**Comment by zhangzujian** [comment]
Could you please attach routes on the node? `ip route show`

**Comment by patriziobassi** [comment]
```
ip route show
default via 10.0.5.1 dev ens3 proto dhcp src 10.0.5.40 metric 100
10.0.5.0/24 dev ens3 proto kernel scope link src 10.0.5.40 metric 100
10.0.5.1 dev ens3 proto dhcp scope link src 10.0.5.40 metric 100
10.0.5.2 dev ens3 proto dhcp scope link src 10.0.5.40 metric 100
10.252.0.0/16 via 100.64.0.1 dev ovn0
100.64.0.0/16 dev ovn0 proto kernel scope link src 100.64.0.7
```

as you can see 10.0.5.40/24, on ens3 interface, is the control plane interface ip used as source in the curl request

10.0.8.0 is not present as it's the underlay net (unnumbered). traffic will go to the physical router which is routing correctly on ens8 interface (where pods are attached). infact from the trace you can see the pod receives the traffic from 10.0.5.40 worker node but when replying that packet is simply lost

**Comment by zhangzujian** [comment]
It's simple to determine whether it's a MTU problem.

If `curl -I <URL>` works but `curl <URL>` does not, packets are dropped because packet length is greater than MTU.

**Comment by patriziobassi** [comment]
It never works when u2o Is enabled

**Comment by patriziobassi** [comment]
@zhangzujian that's great you could reproduce, this is the same behaviour we see.

**Comment by changluyi** [comment]
13. lr_in_policy (northd.c:9244): ip4.dst == $subnet1.u2o_exclude_ip.ip4 && ip4.src == 172.19.0.0/16, priority 31000, uuid ed1cc396
    reg8[0..15] = 0;
    next;
    
    it seems a bug, it should not hit this policy route in my origin design

**Comment by changluyi** [comment]
if your external client ip is not in underlay cidr , it may be ok, you can have a try

**Comment by patriziobassi** [comment]
> if your external client ip is not in underlay cidr , it may be ok, you can have a try

mine is not, and i'm hitting the problem.

**Comment by patriziobassi** [comment]
> 13. lr_in_policy (northd.c:9244): ip4.dst == $subnet1.u2o_exclude_ip.ip4 && ip4.src == 172.19.0.0/16, priority 31000, uuid ed1cc396
>     reg8[0..15] = 0;
>     next;
>     it seems a bug, it should not hit this policy route in my origin design

@changluyi isn't this what policy2 defines, i mean by design according to the comment?

prio 31000 match: "ip4.dst == node ips && ip4.src == underlay subnet cidr"  action: allow 
"policy1 and policy2 allow overlay pod access underlay but when overlay pod access node ip, it should go join subnet,"

this seems the case: the dst is the node ip and the underlay replies.

Just a hint, i'm not expert and i find ovn flow quite difficult to understand: the problem seems that it is designed to go into the join subnet while it should go in the underlay subnet.

**Comment by patriziobassi** [comment]
@changluyi i did a test on my side changing the policy2 match statement like this:

		match2 := fmt.Sprintf("%s.dst != $%s && %s.src == %s", ipSuffix, U2OexcludeIPAs, ipSuffix, cidrBlock)

but doing like that when i create a u2o network the ovs-ovn pods start crashing because they cannot connect upstream server.

do you have any idea how to modify/add the policy rules? i may give a try in my layout but sems @zhangzujian could reproduce too. 
thank you a lot!

**Comment by patriziobassi** [comment]
fantastic job @changluyi, it seems it fixes internal communication and metallb exposure too. of course i need to perform more tests, but looks fine so far. (i manually ported the fix on 1.11.5, replacing the code, i'm not using master).
